util/argo/audit_logger.go

```diff
@@ -0,0 +1,104 @@
+package argo
+
+import (
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"fmt"
+	"time"
+
+	"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+type AuditLogger struct {
+	kIf       kubernetes.Interface
+	component string
+	ns        string
+}
+
+type EventInfo struct {
+	Type   string
+	Reason string
+}
+
+type ObjectRef struct {
+	Name            string
+	Namespace       string
+	ResourceVersion string
+	UID             types.UID
+}
+
+const (
+	EventReasonWorkflowRunning  = "WorkflowRunning"
+	EventReasonWorkflowSucceded = "WorkflowSucceeded"
+	EventReasonWorkflowFailed   = "WorkflowFailed"
+	EventReasonWorkflowTimedOut = "WorkflowTimedOut"
+)
+
+func (l *AuditLogger) logEvent(objMeta ObjectRef, gvk schema.GroupVersionKind, info EventInfo, message string, logFields map[string]string) {
+	logCtx := log.WithFields(log.Fields{
+		"type":   info.Type,
+		"reason": info.Reason,
+	})
+	for field, val := range logFields {
+		logCtx = logCtx.WithField(field, val)
+	}
+
+	switch gvk.Kind {
+	case "Workflow":
+		logCtx = logCtx.WithField("workflow", objMeta.Name)
+	default:
+		logCtx = logCtx.WithField("name", objMeta.Name)
+	}
+	t := metav1.Time{Time: time.Now()}
+	event := v1.Event{
```

Comment:
is this based on existing code? do you have a link if so please?

Reply:
I pretty much copied https://github.com/argoproj/argo-cd/blob/0675ff2fb25fe97856171230f150e6ba5a2563d5/util/argo/audit_logger.go